nesta/packages/arxiv/collect_arxiv.py

```diff
@@ -201,15 +212,221 @@ def arxiv_batch(token, cursor):
+        except (AttributeError, ValueError):  # no matches or strptime conversion fail
+            pass
+    try:
+        return sorted(dates)[-1]
+    except IndexError:
+        raise ValueError("Latest date could not be identified")
+
+
+class BatchedTitles():
+    def __init__(self, ids, batch_size, session):
+        """Extracts batches of titles from the database and yields titles for lookup
+        against the MAG api.
+
+        A lookup dict of {prepared title: [id, ...]} is generated for each batch:
+        self.title_id_lookup
+
+        Args:
+            ids (set): all article ids to be processed
+            batch_size (int): number of ids in each query to the database
+            session (:obj:`sqlalchemy.orm.session`): current session
+
+        Returns:
+            (generator): yields single prepared titles
+        """
+        self.ids = ids
+        self.batch_size = batch_size
+        self.session = session
+        self.title_articles_lookup = defaultdict(list)
+
+    def __getitem__(self, key):
+        """Get articles which match the provided title.
+        Args:
+            key (str): the title of the article
+
+        Returns
+            (:obj:`list` of `str`) article ids which matched the provided title before
+            the request to mag was made.
+        """
+        matching_articles = self.title_articles_lookup.get(key)
+        if matching_articles is None:
+            raise KeyError(f"Title not found in lookup {key}")
+        else:
+            return matching_articles
+
+    def __iter__(self):
+        for batch_of_ids in split_batches(self.ids, self.batch_size):
+            self.title_articles_lookup.clear()
+            for article in (self.session.query(Article)
+                                        .filter(Article.id.in_(batch_of_ids))
+                                        .all()):
+                self.title_articles_lookup[prepare_title(article.title)].append(article.id)
+
+            for title in self.title_articles_lookup:
+                yield title
+
+
+class BatchWriter(list):
+    """A list with functionality to monitor appends.
+    When a specified size is reached a function is called and the list cleared down.
+    """
+    def __init__(self, limit, function, *function_args, **function_kwargs):
+        """
+        Args:
+            limit (int): limit in length triggering a write
+            function (function): the function which will be called
+            args, kwargs: will be passed on to the function call
+        """
+        self.limit = limit
+        self.function = function
+        self.function_args = function_args
+        self.function_kwargs = function_kwargs
+
+    def append(self, item):
+        """Appends to the list and then checks current size against the set limit."""
+        super().append(item)
+        if len(self) >= self.limit:
+            self.write(self.copy())
```

Comment:
what's the purpose of `copy` here?

Reply:
the copy was there to make it testable 😞 if I don't copy it before the list gets cleared my test fails as the values are no longer in the call to my mock. I would love to find a better way to solve this!

Reply:
after changing to your version of the code above it is passing the test fine now 👍